--- ../builder/Dockerfile	2022-10-05 00:59:04.346916757 +0000
+++ Dockerfile	2022-10-05 00:58:30.250652755 +0000
@@ -209,6 +209,13 @@
   && chmod +x /usr/local/bin/autouseradd \
   && rm autouseradd.tar.gz
 
+# see README.md for instructions on how to install Antithesis instrumentation library.
+
+RUN mkdir -p /opt/antithesis/bin/ && mkdir -p /opt/antithesis/lib/ && mkdir -p /opt/antithesis/instrumentation/go/wrappers/
+COPY libvoidstar.so /opt/antithesis/lib/
+COPY goinstrumentor /opt/antithesis/bin/
+COPY go.mod wrappers.go /opt/antithesis/instrumentation/go/wrappers/
+
 COPY entrypoint.sh /usr/local/bin
 
 ENTRYPOINT ["autouseradd", "--user", "roach", "--no-create-home", "--", "entrypoint.sh"]
