diff -urN a/errcheck/analyzer.go b/errcheck/analyzer.go
--- a/errcheck/analyzer.go
+++ b/errcheck/analyzer.go
@@ -5,6 +5,7 @@
 	"go/ast"
 	"reflect"
 	"regexp"
+	"strings"
 
 	"golang.org/x/tools/go/analysis"
 )
@@ -20,6 +21,7 @@
 	argBlank       bool
 	argAsserts     bool
 	argExcludeFile string
+	argExcludes    string
 	argExcludeOnly bool
 )
 
@@ -27,6 +29,7 @@
 	Analyzer.Flags.BoolVar(&argBlank, "blank", false, "if true, check for errors assigned to blank identifier")
 	Analyzer.Flags.BoolVar(&argAsserts, "assert", false, "if true, check for ignored type assertion results")
 	Analyzer.Flags.StringVar(&argExcludeFile, "exclude", "", "Path to a file containing a list of functions to exclude from checking")
+	Analyzer.Flags.StringVar(&argExcludes, "excludes", "", "Contents of the exclude file as a string (overrides -exclude)")
 	Analyzer.Flags.BoolVar(&argExcludeOnly, "excludeonly", false, "Use only excludes from exclude file")
 }
 
@@ -37,7 +40,14 @@
 			exclude[name] = true
 		}
 	}
-	if argExcludeFile != "" {
+	if argExcludes != "" {
+		for _, name := range strings.Split(argExcludes, "\n") {
+			if strings.HasPrefix(name, "//") || name == "" {
+				continue
+			}
+			exclude[name] = true
+		}
+	} else if argExcludeFile != "" {
 		excludes, err := ReadExcludes(argExcludeFile)
 		if err != nil {
 			return nil, fmt.Errorf("Could not read exclude file: %v\n", err)
@@ -63,8 +73,9 @@
 		ast.Walk(v, f)
 
 		for _, err := range v.errors {
+			fsetFile := pass.Fset.File(f.Pos())
 			pass.Report(analysis.Diagnostic{
-				Pos:      pass.Fset.File(f.Pos()).Pos(err.Pos.Offset),
+				Pos:      fsetFile.Pos(err.Pos.Offset),
 				Message:  "unchecked error",
 				Category: "errcheck",
 			})
