diff --git a/internal/buildtags/BUILD.bazel b/internal/buildtags/BUILD.bazel
--- a/internal/buildtags/BUILD.bazel
+++ b/internal/buildtags/BUILD.bazel
@@ -6,6 +6,7 @@ go_library(
         "cgo_off.go",
         "cgo_on.go",
         "invariants_off.go",
+        "invariants_on.go",
         "race_off.go",
         "race_on.go",
         "tracing_off.go",
diff --git a/objstorage/objstorageprovider/objiotracing/BUILD.bazel b/objstorage/objstorageprovider/objiotracing/BUILD.bazel
--- a/objstorage/objstorageprovider/objiotracing/BUILD.bazel
+++ b/objstorage/objstorageprovider/objiotracing/BUILD.bazel
@@ -5,6 +5,7 @@ go_library(
     srcs = [
         "obj_io_tracing.go",
         "obj_io_tracing_off.go",
+        "obj_io_tracing_on.go",
     ],
     importpath = "github.com/cockroachdb/pebble/objstorage/objstorageprovider/objiotracing",
     visibility = ["//visibility:public"],
