diff -urN a/internal/facts/BUILD.bazel b/internal/facts/BUILD.bazel
--- a/internal/facts/BUILD.bazel
+++ b/internal/facts/BUILD.bazel
@@ -7,7 +7,7 @@
         "imports.go",
     ],
     importpath = "golang.org/x/tools/internal/facts",
-    visibility = ["//:__subpackages__"],
+    visibility = ["//visibility:public"],
     deps = [
         "//go/analysis",
         "//go/types/objectpath",
@@ -18,7 +18,7 @@
 alias(
     name = "go_default_library",
     actual = ":facts",
-    visibility = ["//:__subpackages__"],
+    visibility = ["//visibility:public"],
 )
 
 go_test(
