commit 941ed7af9e23f6b2629deef4c472e723b284276e
Author: Tamir Duberstein <tamird@gmail.com>
Date:   Thu Apr 13 20:09:41 2017 -0400

    CMake: add support for SSE4.2

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 924ec711..cdf9feef 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -113,6 +113,11 @@ if(WIN32)
       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mavx2")
     endif()
   endif()
+else()
+  option(WITH_SSE42 "build with SSE4.2" ON)
+  if(WITH_SSE42)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse4.2")
+  endif()
 endif()

 set(BUILD_VERSION_CC ${CMAKE_BINARY_DIR}/build_version.cc)
