diff --git a/db/c_test.c b/db/c_test.c
index f30b5275..6b727d7e 100644
--- a/db/c_test.c
+++ b/db/c_test.c
@@ -20,7 +20,7 @@
 // Can not use port/port.h macros as this is a c file
 #ifdef OS_WIN
 
-#include <Windows.h>
+#include <windows.h>
 
 #define snprintf _snprintf
 
diff --git a/port/win/env_win.cc b/port/win/env_win.cc
index ced56f7f..5f079c4b 100644
--- a/port/win/env_win.cc
+++ b/port/win/env_win.cc
@@ -33,8 +33,8 @@
 #include "util/thread_status_updater.h"
 #include "util/thread_status_util.h"
 
-#include <Rpc.h>  // For UUID generation
-#include <Windows.h>
+#include <rpc.h>  // For UUID generation
+#include <windows.h>
 
 namespace rocksdb {
 
diff --git a/port/win/env_win.h b/port/win/env_win.h
index e0e283af..75164c23 100644
--- a/port/win/env_win.h
+++ b/port/win/env_win.h
@@ -21,7 +21,7 @@
 #include "util/threadpool_imp.h"
 
 #include <stdint.h>
-#include <Windows.h>
+#include <windows.h>
 
 #include <mutex>
 #include <vector>
diff --git a/port/win/io_win.h b/port/win/io_win.h
index a434ab77..f50c237a 100644
--- a/port/win/io_win.h
+++ b/port/win/io_win.h
@@ -16,7 +16,7 @@
 #include "rocksdb/env.h"
 #include "util/aligned_buffer.h"
 
-#include <Windows.h>
+#include <windows.h>
 
 
 namespace rocksdb {
diff --git a/port/win/win_logger.h b/port/win/win_logger.h
index 84971363..52913b97 100644
--- a/port/win/win_logger.h
+++ b/port/win/win_logger.h
@@ -17,7 +17,7 @@
 #include "rocksdb/env.h"
 
 #include <stdint.h>
-#include <Windows.h>
+#include <windows.h>
 
 namespace rocksdb {
 
diff --git a/port/win/win_thread.cc b/port/win/win_thread.cc
index 987950a6..cda0f2e8 100644
--- a/port/win/win_thread.cc
+++ b/port/win/win_thread.cc
@@ -11,7 +11,7 @@
 
 #include <assert.h>
 #include <process.h> // __beginthreadex
-#include <Windows.h>
+#include <windows.h>
 
 #include <stdexcept>
 #include <system_error>
diff --git a/port/win/xpress_win.cc b/port/win/xpress_win.cc
index a0206b57..f36a4522 100644
--- a/port/win/xpress_win.cc
+++ b/port/win/xpress_win.cc
@@ -8,7 +8,7 @@
 // found in the LICENSE file. See the AUTHORS file for names of contributors.
 
 #include "port/win/xpress_win.h"
-#include <Windows.h>
+#include <windows.h>
 
 #include <cassert>
 #include <memory>
