diff --git a/scripts/setup-go-submodule.sh b/scripts/setup-go-submodule.sh
index 151bfa4feb..d4796da179 100755
--- a/scripts/setup-go-submodule.sh
+++ b/scripts/setup-go-submodule.sh
@@ -14,7 +14,7 @@ if [ -z "${GIT_REF}" ]; then
     fi
 fi
 
-git submodule add --force https://github.com/golang/go.git
+git submodule add --force ${GOLANG_REPO:-https://github.com/golang/go.git}
 git submodule update
 
 pushd go
