diff --git a/pkg/ccl/storageccl/export_storage_test.go b/pkg/ccl/storageccl/export_storage_test.go
index c71a1cd1e..908d5a6db 100644
--- a/pkg/ccl/storageccl/export_storage_test.go
+++ b/pkg/ccl/storageccl/export_storage_test.go
@@ -150,7 +150,6 @@ func TestPutS3(t *testing.T) {
 	}
 	bucket := os.Getenv("AWS_S3_BUCKET")
 	if bucket == "" {
-		t.Skip("AWS_S3_BUCKET env var must be set")
 	}
 
 	// TODO(dt): this prevents leaking an http conn goroutine.
